### DISCUSSION TITLE
FIP Proposal: Ethereum State on Filecoin

### DISCUSSION THREAD

**Discussion by nasdf** [filecoin-project/FIPs#270] [Enhancements - Technical]
# FIP Proposal: Ethereum State on Filecoin

Filecoin has the potential to store state from other blockchains. With [EIP-4444](https://eips.ethereum.org/EIPS/eip-4444) on the horizon, we propose a method to use Filecoin to store historical Ethereum state. This enables novel use cases of Filecoin as a Layer 0, while simultaneously laying the groundwork for an IPFS-powered Ethereum light client.

## High-level ideas

- Store Ethereum state on Filecoin
    - Ethereum state is bridged to IPFS as the hot layer
    - Historical state is archived to Filecoin as the cold layer
- Access Ethereum state from the FVM. 
    - Rolling up Ethereum state onto Filecoin as the layer 0.
    - Enables cross-chain, trustless bridges from the FVM.
    - New methods for token swapping, metadata sharing, cross-chain state sharing.
- Access Ethereum state from IPFS.
    - IPFS plugin provides a decentralized Ethereum RPC.
    - fetch and validate any piece of state in a distributed way.
    - Decentralized, historical archive data.

## Milestones

- Support for Ethereum IPLD codecs included in IPFS & Filecoin clients
- Tools for bridging and creating storage deals for historical state
- Support for adding an Ethereum RPC to IPFS via plugin
- In browser Ethereum light clients powered by IPFS

## Current Questions

- Can we access IPLD objects that we store on Filecoin from the FVM?
- Can we access data stored in a deal from the FVM?
- Can the FVM create or update storage deals?

## Prior Work

https://github.com/vulcanize/go-codec-dageth
https://github.com/valist-io/leo

## Authors

@nasdf
@awantoch
@jiyuu-jin

**Comment by kaitlin-beegle** [comment]
Hi @nasdf, thanks for taking the time to submit this proposal! I'm Kaitlin, and I coordinate protocol governance on behalf of the Filecoin Foundation. 

I will be sure to flag your post in this week's governance update, and am always available to answer any questions you may have about the FIPs process!

Feel free to reach out to me on Github or on Filecoin Slack (@kaitlin_ff), or [put some time on my calendar](https://calendly.com/kaitlin-45/30min-1?month=2022-01) for us to chat. Cheers!

**Comment by anorth** [comment]
Thanks @nasdf! To briefly answer a couple of your question as the network exists today and the near future:

> Can we access IPLD objects that we store on Filecoin from the FVM?

IPLD data stored in the Filecoin blockchain state can be accessed from contracts running on the FVM, yes. But this is the relatively expensive consensus state, and I'm not sure if that's what you meant. It'll be appropriate for metadata, but not large Eth state blobs. If you meant IPLD data stored in Filecoin deals, see below.

> Can we access data stored in a deal from the FVM?

Not directly. Deal data is generally stored on only one node, so cannot be accessed by most block producers or validators, thus cannot be accessed by smart contracts (except for the CID that is the root of a deal's data). However, there are a range of possible mechanisms by which we could enable off-chain access to and computation over the data stored in a deal, with the results of that computation stored in another deal (or a hash of the result stored on-chain). These would be asynchronous, market-based computations, but we're quite optimistic about building up the primitives needed for a robust computation market. Our name for this general effort is "compute over data".

> Can the FVM create or update storage deals?

We will make it so.

**Reply by nasdf** [reply]
Hi @anorth,

Thank you for clearing up those questions.  I think most of the ideas in the proposal are still possible with some of it moving to off-chain computation and verification in the FVM.

> These would be asynchronous, market-based computations, but we're quite optimistic about building up the primitives needed for a robust computation market.

This is such an awesome concept that deserves more attention. Consensus protocols as smart contracts with automated payouts.
